feat: add lyrics window

tisfeng opened the pull request:
Close https://github.com/kartik-venugopal/aural-player/issues/111

tisfeng committed:
feat: add lyrics scroll view

tisfeng committed:
Merge branch 'test-lyrics' into feat-lyrics

tisfeng committed:
fix: solve merge error

tisfeng committed:
feat: add an option for showing Chinese lyrics transaltion

tisfeng committed:
fix: close old lyrics search window when starting search

kartik-venugopal commented:
Thanks @tisfeng . I took a brief look ... this is a fairly large changelist, so it will take me some time to review, test, and merge everything. I am currently working on another new feature, so I'll try to get to this within the next few days.

kartik-venugopal commented:
Hi @tisfeng , thanks for your work on this. But I will be honest - the scope of these changes is much larger than what we discussed in issue number 111.

At this time, I do not want to add online searching of lyrics to Aural, I only want to add local file integration and embedded (dynamic)lyrics support.

I also do not want to add a large number of package dependencies to Aural. This is going to add too much complexity for little benefit.

Lastly, I do not want to add Chinese translations to Aural. Localization / translation is something that would need to be more language-agnostic.

Are you able to factor out only the local lyrics support and displaying of dynamic lyrics, without adding any package dependencies ? Thanks.

tisfeng commented:
I understand your thoughts.

I really like lyrics, but manually adding lyrics for each song is too troublesome. I believe other users who love lyrics have similar thoughts, so I added an online lyrics search feature.

As for the lyrics translation, it is provided by a Chinese online lyrics service QQ, which includes Chinese translations along with the lyrics, because it is very helpful for Chinese users, so I added it as well.

For these two features, if you don't want them, I can remove them.

tisfeng commented:
Regarding the package dependencies, I actually only added one, [LyricsXCore](https://github.com/ddddxxx/LyricsXCore), which is a library extracted from the well-known lyrics project [LyricsX](https://github.com/ddddxxx/LyricsX). It has been tested over time, and it is reliable.

The handling of lrcx lyrics is quite complex. [LyricsCore](https://github.com/ddddxxx/LyricsKit/tree/master/Sources/LyricsCore), as an excellent third-party library in this field, is a wise choice if we want Aural to support lrcx format lyrics well.

I forked [LyricsXCore](https://github.com/tisfeng/LyricsXCore), fixed some minor issues, and added support for [karaoke lyrics effects](https://github.com/tisfeng/LyricsXCore/blob/dev/Sources/LyricsUI/KaraokeLyricsView.swift).

If you have any other suggestions, please let me know.

<img width="1671" alt="image" src="https://github.com/user-attachments/assets/7887c60d-f5ab-47f5-82ae-f7c0ee66bf9e" />

kartik-venugopal commented:
I see your point. Thanks for explaining your changes. I agree, it would be nice to have search capabilities. Let's try to add it, then.

I was referring to these dependencies (I guess these are transitive dependencies used by LyricsX).

![Screenshot 2024-12-31 at 9 36 43 AM](https://github.com/user-attachments/assets/c09027ea-94c9-449e-9668-9103973266cb)

----------------------------

So, can we just do this:
1 - take out the Chinese translation part
2 - I'm getting a compilation error from your project (I downloaded your code as Zip and just tried to build). I'm guessing I need a newer Xcode / Swift version ? I'm on Xcode 15.3 

![Screenshot 2024-12-31 at 9 37 17 AM](https://github.com/user-attachments/assets/359623be-4eef-4fd1-8d1b-f067687d8a09)

kartik-venugopal commented:
EDIT - I downloaded Xcode 16.2 (Swift 6) and I was able to build and run. Will play with it in the next few days!

tisfeng commented:
> 1 - take out the Chinese translation part

Sorry, my previous description was inaccurate. In fact, this option is used to display the translation of the song lyrics, if available.

It is not limited to Chinese lyrics translation; for example, if a song is in English and the user has prepared a French translation of the lyrics, as long as the lyrics file is in lrcx format, it can be displayed correctly in the Aural lyrics window.

I can change this label in the settings page if you confirm that you can accept this option.

<img width="746" alt="image" src="https://github.com/user-attachments/assets/e6c7ca41-72e9-4eca-824f-a027427c5bcc" />

kartik-venugopal commented:
Ok, sounds good. Please change the preference setting (and all variables in the code, etc) to reflect the fact that it is language-agnostic. And I think we're good to merge at that point!

Thanks!

tisfeng commented:
>I was referring to these dependencies (I guess these are transitive dependencies used by LyricsX).

Yes, initially I also didn't expect LyricsX to use so many dependencies internally.

I also agree that we shouldn't introduce too many dependencies, as this increases the complexity of the program and can make it difficult to maintain.

However, after careful review, I found that there are two reasons why it includes so many dependencies:

1. [LyricsXCore](https://github.com/ddddxxx/LyricsXCore) is a very powerful library, it even supports Linux, and to achieve this, it added some cross-platform open-source libraries from [CombineX](https://github.com/cx-org).
2. It uses a program architecture called [The Composable Architecture (TCA, for short)](https://github.com/pointfreeco/swift-composable-architecture), which includes many internal dependencies. It's a relatively popular architecture library in Swift, and is considered quite reliable.

Considering the powerful functionality and stability of LyricsX, I think perhaps we can accept it.

tisfeng committed:
fix: rename to showLyricsTranslation
